File: slycalendarview/src/main/java/ru/slybeaver/slycalendarview/GridAdapter.java
package ru.slybeaver.slycalendarview;

import android.content.Context;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.LayerDrawable;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

import ru.slybeaver.slycalendarview.listeners.DateSelectListener;
import ru.slybeaver.slycalendarview.listeners.GridChangeListener;

/**
 * Created by psinetron on 29/11/2018.
 * http://slybeaver.ru
 */
public class GridAdapter extends ArrayAdapter {

    private SlyCalendarData calendarData;
    private int shiftMonth;
    private ArrayList<Date> monthlyDates = new ArrayList<>();
    private DateSelectListener listener;
    private LayoutInflater inflater;
    private GridChangeListener gridListener;


    public GridAdapter(
            @NonNull Context context,
            @NonNull SlyCalendarData calendarData,
            int shiftMonth,
            @Nullable DateSelectListener listener,
            GridChangeListener gridListener
    ) {
        super(context, R.layout.slycalendar_single_cell);
        this.calendarData = calendarData;
        this.listener = listener;
        this.inflater = LayoutInflater.from(context);
        this.shiftMonth = shiftMonth;
        this.gridListener = gridListener;
        init();
    }

    @NonNull
    @Override
    public View getView(final int position, @Nullable View convertView, @Nullable ViewGroup parent) {
        Calendar dateCal = Calendar.getInstance();
        dateCal.setTime(monthlyDates.get(position));

        Calendar calendarStart = createCalendarWith(calendarData.getSelectedStartDate());
        Calendar calendarEnd = createCalendarWith(calendarData.getSelectedEndDate());


        View view = convertView;
        if (view == null) {
            view = inflater.inflate(R.layout.slycalendar_single_cell, parent, false);
        }
        View cell = view.findViewById(R.id.cellView);
        TextView textDate = view.findViewById(R.id.txtDate);
        View frame = view.findViewById(R.id.frameSelected);


        textDate.setText(String.valueOf(dateCal.get(Calendar.DAY_OF_MONTH)));
        cell.setBackgroundResource(R.color.slycalendar_defBackgroundColor);


        cell.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Calendar selectedDate = Calendar.getInstance();
                selectedDate.setTime(monthlyDates.get(position));
                selectedDate.set(Calendar.HOUR, 0);
                selectedDate.set(Calendar.MINUTE, 0);
                selectedDate.set(Calendar.SECOND, 0);
                selectedDate.set(Calendar.MILLISECOND, 0);
                if (listener != null && isSelected(selectedDate))
                    listener.dateSelect(selectedDate.getTime());
                notifyDataSetChanged();
                gridListener.gridChanged();
            }
        });

        cell.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                Calendar selectedDate = Calendar.getInstance();
                selectedDate.setTime(monthlyDates.get(position));
                if (listener != null && isSelected(selectedDate))
                    listener.dateLongSelect(selectedDate.getTime());
                notifyDataSetChanged();
                gridListener.gridChanged();
                return true;
            }
        });


        cell.setBackgroundColor(calendarData.getBackgroundColor());

        if (calendarStart != null && calendarEnd != null) {
            if (dateCal.get(Calendar.DAY_OF_YEAR) == calendarStart.get(Calendar.DAY_OF_YEAR) &&
                    dateCal.get(Calendar.YEAR) == calendarStart.get(Calendar.YEAR)) {

                setUpShapeBackgroundOn(view, R.drawable.slycalendar_start_day);

            } else if (dateCal.getTimeInMillis() > calendarStart.getTimeInMillis()
                    && dateCal.getTimeInMillis() < calendarEnd.getTimeInMillis()) {

                setUpShapeBackgroundOn(view, R.drawable.slycalendar_middle_day);

                if (isMonday(position)) {

                    setUpShapeBackgroundOn(view, R.drawable.slycalendar_start_day);
                }
                if (nextDayIsMonday(position)) {

                    setUpShapeBackgroundOn(view, R.drawable.slycalendar_end_day);
                }


            } else if (dateCal.get(Calendar.DAY_OF_YEAR) == calendarEnd.get(Calendar.DAY_OF_YEAR)
                    && dateCal.get(Calendar.YEAR) == calendarEnd.get(Calendar.YEAR)) {

                setUpShapeBackgroundOn(view, R.drawable.slycalendar_end_day);
            }
        }


        Calendar currentDate = Calendar.getInstance();


        frame.setBackgroundResource(0);
        textDate.setTextColor(calendarData.getTextColor());
        if (calendarStart != null && dateCal.get(Calendar.DAY_OF_YEAR) == calendarStart.get(Calendar.DAY_OF_YEAR) && calendarStart.get(Calendar.YEAR) == dateCal.get(Calendar.YEAR)) {
            LayerDrawable shape = (LayerDrawable) ContextCompat.getDrawable(getContext(), R.drawable.slycalendar_selected_day);
            assert shape != null;
            ((GradientDrawable) shape.findDrawableByLayerId(R.id.selectedDateShapeItem)).setColor(calendarData.getSelectedColor());
            frame.setBackground(shape);
            textDate.setTextColor(calendarData.getSelectedTextColor());
            textDate.setTypeface(Typeface.create("sans-serif-medium", Typeface.NORMAL));
        }

        if (calendarEnd != null && dateCal.get(Calendar.DAY_OF_YEAR) == calendarEnd.get(Calendar.DAY_OF_YEAR) && calendarEnd.get(Calendar.YEAR) == dateCal.get(Calendar.YEAR)) {
            LayerDrawable shape = (LayerDrawable) ContextCompat.getDrawable(getContext(), R.drawable.slycalendar_selected_day);
            assert shape != null;
            ((GradientDrawable) shape.findDrawableByLayerId(R.id.selectedDateShapeItem)).setColor(calendarData.getSelectedColor());
            frame.setBackground(shape);
            textDate.setTextColor(calendarData.getSelectedTextColor());
            textDate.setTypeface(Typeface.create("sans-serif-medium", Typeface.NORMAL));

        }

        if (isBefore(currentDate, dateCal)) {
            textDate.setAlpha(1);
        } else {
            textDate.setAlpha(.2f);
        }

        return view;
    }


    private boolean isBefore(Calendar current, Calendar another) {
        return current.getTimeInMillis() > another.getTimeInMillis();
    }


    private boolean isSelected(Calendar selectedDate) {
        Calendar today = Calendar.getInstance(Locale.getDefault());
        return isBefore(today, selectedDate);
    }

    private boolean isMonday(int position) {
        return position % 7 == 0;
    }

    private boolean nextDayIsMonday(int position) {
        return (position + 1) % 7 == 0;
    }

    private Calendar createCalendarWith(Date selectedStartDate) {
        Calendar calendarStart = null;
        if (selectedStartDate != null) {
            calendarStart = Calendar.getInstance();
            calendarStart.setTime(selectedStartDate);
            calendarStart.set(Calendar.HOUR, 0);
            calendarStart.set(Calendar.MINUTE, 0);
            calendarStart.set(Calendar.SECOND, 0);
            calendarStart.set(Calendar.MILLISECOND, 0);
        }
        return calendarStart;
    }

    private void setUpShapeBackgroundOn(View cell, @DrawableRes int drawableResource) {
        LayerDrawable shape = (LayerDrawable) ContextCompat.getDrawable(getContext(), drawableResource);
        assert shape != null;
        ((GradientDrawable) shape.findDrawableByLayerId(R.id.dateShapeItem)).setColor(calendarData.getSelectedColor());
        (shape.findDrawableByLayerId(R.id.dateShapeItem)).setAlpha(20);
        cell.setBackground(shape);

    }

    @Override
    public int getCount() {
        return monthlyDates.size();
    }

    @Nullable
    @Override
    public Date getItem(int position) {
        return monthlyDates.get(position);
    }

    private void init() {
        monthlyDates = new ArrayList<>();
        Calendar calendar = Calendar.getInstance(Locale.getDefault());
        calendar.setTime(calendarData.getShowDate());
        calendar.add(Calendar.MONTH, shiftMonth);
        calendar.set(Calendar.DAY_OF_MONTH, 1);
        int firstDayOfTheMonth = calendarData.isFirstMonday() ? calendar.get(Calendar.DAY_OF_WEEK) - 2 : calendar.get(Calendar.DAY_OF_WEEK) - 1;
        calendar.add(Calendar.DAY_OF_MONTH, -firstDayOfTheMonth);
        int MAX_CALENDAR_COLUMN = 42;
        while (monthlyDates.size() < MAX_CALENDAR_COLUMN) {
            monthlyDates.add(calendar.getTime());
            calendar.add(Calendar.DAY_OF_MONTH, 1);
        }
    }


}

File: slycalendarview/src/main/java/ru/slybeaver/slycalendarview/GridAdapter.kt
package ru.slybeaver.slycalendarview

import android.content.Context
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.LayerDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.TextView
import androidx.core.content.ContextCompat
import ru.slybeaver.slycalendarview.listeners.DateSelectListener
import ru.slybeaver.slycalendarview.listeners.GridChangeListener
import java.util.*

/**
 * Created by psinetron on 27/05/20
 * http://slybeaver.ru
 */
class GridAdapter(
        context: Context,
        calendarData: SlyCalendarData,
        shiftMonth: Int,
        listener: DateSelectListener?,
        gridListener: GridChangeListener?
) : BaseAdapter() {
    private var context: Context? = null
    private var calendarData: SlyCalendarData? = null
    private var listener: DateSelectListener? = null
    private var gridListener: GridChangeListener? = null
    private var shiftMonth = 0

    private var monthlyDates = ArrayList<Date>()
    private var inflater: LayoutInflater? = null

    init {
        this.context = context
        this.calendarData = calendarData
        this.listener = listener
        this.gridListener = gridListener
        this.shiftMonth = shiftMonth
        this.inflater = LayoutInflater.from(context)

        monthlyDates = ArrayList()
        val calendar = Calendar.getInstance(Locale.getDefault())
        calendar.time = calendarData.showDate!!
        calendar.add(Calendar.MONTH, shiftMonth)
        calendar[Calendar.DAY_OF_MONTH] = 1
        val firstDayOfTheMonth = if (calendarData.firstMonday) calendar[Calendar.DAY_OF_WEEK] - 2 else calendar[Calendar.DAY_OF_WEEK] - 1
        calendar.add(Calendar.DAY_OF_MONTH, -firstDayOfTheMonth)
        val MAX_CALENDAR_COLUMN = 42
        while (monthlyDates.size < MAX_CALENDAR_COLUMN) {
            monthlyDates.add(calendar.time)
            calendar.add(Calendar.DAY_OF_MONTH, 1)
        }
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup?): View {
        val dateCal = Calendar.getInstance()
        dateCal.time = monthlyDates[position]
        var calendarStart: Calendar? = null
        if (calendarData?.selectedStartDate != null) {
            calendarStart = Calendar.getInstance()
            calendarStart.time = calendarData?.selectedStartDate!!
            calendarStart[Calendar.HOUR] = 0
            calendarStart[Calendar.MINUTE] = 0
            calendarStart[Calendar.SECOND] = 0
            calendarStart[Calendar.MILLISECOND] = 0
        }
        var calendarEnd: Calendar? = null
        if (calendarData?.selectedEndDate != null) {
            calendarEnd = Calendar.getInstance()
            calendarEnd.time = calendarData?.selectedEndDate!!
            calendarEnd[Calendar.HOUR] = 0
            calendarEnd[Calendar.MINUTE] = 0
            calendarEnd[Calendar.SECOND] = 0
            calendarEnd[Calendar.MILLISECOND] = 0
        }
        var view = convertView
        if (view == null) {
            view = inflater!!.inflate(R.layout.slycalendar_single_cell, parent, false)
        }
        (view!!.findViewById<View>(R.id.txtDate) as TextView).text = dateCal[Calendar.DAY_OF_MONTH].toString()
        view.findViewById<View>(R.id.cellView).setBackgroundResource(R.color.slycalendar_defBackgroundColor)
        view.findViewById<View>(R.id.cellView).setOnClickListener {
            val selectedDate = Calendar.getInstance()
            selectedDate.time = monthlyDates[position]
            selectedDate[Calendar.HOUR] = 0
            selectedDate[Calendar.MINUTE] = 0
            selectedDate[Calendar.SECOND] = 0
            selectedDate[Calendar.MILLISECOND] = 0
            listener?.dateSelect(selectedDate.time)
            notifyDataSetChanged()
            gridListener!!.gridChanged()
        }
        view.findViewById<View>(R.id.cellView).setOnLongClickListener {
            val selectedDate = Calendar.getInstance()
            selectedDate.time = monthlyDates[position]
            selectedDate[Calendar.HOUR] = 0
            selectedDate[Calendar.MINUTE] = 0
            selectedDate[Calendar.SECOND] = 0
            selectedDate[Calendar.MILLISECOND] = 0
            listener?.dateLongSelect(monthlyDates[position])
            notifyDataSetChanged()
            gridListener!!.gridChanged()
            true
        }
        view.findViewById<View>(R.id.cellView).setBackgroundColor(calendarData?.backgroundColor!!)
        if (calendarStart != null && calendarEnd != null) {
            if (dateCal[Calendar.DAY_OF_YEAR] == calendarStart[Calendar.DAY_OF_YEAR] && dateCal[Calendar.YEAR] == calendarStart[Calendar.YEAR]) {
                val shape = (ContextCompat.getDrawable(context!!, R.drawable.slycalendar_start_day) as LayerDrawable?)!!
                (shape.findDrawableByLayerId(R.id.dateShapeItem) as GradientDrawable).setColor(calendarData?.selectedColor!!)
                shape.findDrawableByLayerId(R.id.dateShapeItem).alpha = 20
                view.findViewById<View>(R.id.cellView).background = shape
            } else if (dateCal.timeInMillis > calendarStart.timeInMillis && dateCal.timeInMillis < calendarEnd.timeInMillis) {
                var shape = (ContextCompat.getDrawable(context!!, R.drawable.slycalendar_middle_day) as LayerDrawable?)!!
                (shape.findDrawableByLayerId(R.id.dateShapeItem) as GradientDrawable).setColor(calendarData?.selectedColor!!)
                shape.findDrawableByLayerId(R.id.dateShapeItem).alpha = 20
                view.findViewById<View>(R.id.cellView).background = shape
                if (position % 7 == 0) {
                    shape = ContextCompat.getDrawable(context!!, R.drawable.slycalendar_start_day) as LayerDrawable
                    assert(shape != null)
                    (shape.findDrawableByLayerId(R.id.dateShapeItem) as GradientDrawable).setColor(calendarData?.selectedColor!!)
                    shape.findDrawableByLayerId(R.id.dateShapeItem).alpha = 20
                    view.findViewById<View>(R.id.cellView).background = shape
                }
                if ((position + 1) % 7 == 0) {
                    shape = ContextCompat.getDrawable(context!!, R.drawable.slycalendar_end_day) as LayerDrawable
                    assert(shape != null)
                    (shape.findDrawableByLayerId(R.id.dateShapeItem) as GradientDrawable).setColor(calendarData?.selectedColor!!)
                    shape.findDrawableByLayerId(R.id.dateShapeItem).alpha = 20
                    view.findViewById<View>(R.id.cellView).background = shape
                }
            } else if (dateCal[Calendar.DAY_OF_YEAR] == calendarEnd[Calendar.DAY_OF_YEAR] && dateCal[Calendar.YEAR] == calendarEnd[Calendar.YEAR]) {
                val shape = (ContextCompat.getDrawable(context!!, R.drawable.slycalendar_end_day) as LayerDrawable?)!!
                (shape.findDrawableByLayerId(R.id.dateShapeItem) as GradientDrawable).setColor(calendarData?.selectedColor!!)
                shape.findDrawableByLayerId(R.id.dateShapeItem).alpha = 20
                view.findViewById<View>(R.id.cellView).background = shape
            }
        }
        val currentDate = Calendar.getInstance()
        currentDate.time = calendarData?.showDate!!
        currentDate.add(Calendar.MONTH, shiftMonth)
        view.findViewById<View>(R.id.frameSelected).setBackgroundResource(0)
        (view.findViewById<View>(R.id.txtDate) as TextView).setTextColor(calendarData?.textColor!!)
        if (calendarStart != null && dateCal[Calendar.DAY_OF_YEAR] == calendarStart[Calendar.DAY_OF_YEAR] && currentDate[Calendar.MONTH] == dateCal[Calendar.MONTH] && dateCal[Calendar.YEAR] == calendarStart[Calendar.YEAR]) {
            val shape = (ContextCompat.getDrawable(context!!, R.drawable.slycalendar_selected_day) as LayerDrawable?)!!
            (shape.findDrawableByLayerId(R.id.selectedDateShapeItem) as GradientDrawable).setColor(calendarData?.selectedColor!!)
            view.findViewById<View>(R.id.frameSelected).background = shape
            (view.findViewById<View>(R.id.txtDate) as TextView).setTextColor(calendarData?.selectedTextColor!!)
            (view.findViewById<View>(R.id.txtDate) as TextView).typeface = Typeface.create("sans-serif-medium", Typeface.NORMAL)
        }
        if (calendarEnd != null && dateCal[Calendar.DAY_OF_YEAR] == calendarEnd[Calendar.DAY_OF_YEAR] && currentDate[Calendar.MONTH] == dateCal[Calendar.MONTH] && dateCal[Calendar.YEAR] == calendarEnd[Calendar.YEAR]) {
            val shape = (ContextCompat.getDrawable(context!!, R.drawable.slycalendar_selected_day) as LayerDrawable?)!!
            (shape.findDrawableByLayerId(R.id.selectedDateShapeItem) as GradientDrawable).setColor(calendarData?.selectedColor!!)
            view.findViewById<View>(R.id.frameSelected).background = shape
            (view.findViewById<View>(R.id.txtDate) as TextView).setTextColor(calendarData?.selectedTextColor!!)
        }
        if (currentDate[Calendar.MONTH] == dateCal[Calendar.MONTH]) {
            view.findViewById<View>(R.id.txtDate).alpha = 1f
        } else {
            view.findViewById<View>(R.id.txtDate).alpha = .2f
        }
        return view
    }

    override fun getCount(): Int {
        return monthlyDates.size
    }

    override fun getItem(position: Int): Date? {
        return monthlyDates[position]
    }

    override fun getItemId(position: Int): Long {
        return monthlyDates[position].time
    }
}